src/frequenz/dispatch/_bg_service.py

```diff
@@ -396,23 +396,30 @@ async def _fetch(self, timer: Timer) -> None:
                         )
                         await self._lifecycle_events_tx.send(Updated(dispatch=dispatch))
 
-            _logger.debug("Received %s dispatches", len(self._dispatches))
+            _logger.debug("Received %s dispatches", len(new_dispatches))
 
         except grpc.aio.AioRpcError as error:
             _logger.error("Error fetching dispatches: %s", error)
-            self._dispatches = old_dispatches
             return
 
-        for dispatch in old_dispatches.values():
+        # Delete old dispatches
+        for dispatch_id in old_dispatches:
+            if dispatch_id in new_dispatches:
+                continue
+
+            dispatch = self._dispatches.pop(dispatch_id)
```

Comment:
`pop()` will raise if the key is not found, not sure if this could happen if you are dealing with some dict inconsistencies here. If it could happen you should either first check if it exists or catch the `KeyError` `pop` could raise.